tools/clang/lib/SPIRV/EmitSPIRVAction.cpp

```diff
@@ -694,30 +716,132 @@ class SPIRVEmitter : public ASTConsumer {
+    // Under such cases, we need to create a new vector swizzling involving
+    // both the lhs and rhs vectors and then write the result of this swizzling
+    // into the base vector of lhs.
+    // For example, for vec4.yz = vec2, we nee to do the following:
+    //
+    //   %vec4Val = OpLoad %v4float %vec4
+    //   %vec2Val = OpLoad %v2float %vec2
+    //   %shuffle = OpVectorShuffle %v4float %vec4Val %vec2Val 0 4 5 3
+    //   OpStore %vec4 %shuffle
+    //
+    // When doing the vector shuffle, we use the lhs base vector as the first
+    // vector and the rhs vector as the second vector. We select all elements
+    // in the second vector and all elements untouched in the first vector.
 
-    // Handle assignment first since we need to evaluate rhs before lhs.
-    // For other binary operations, we need to evaluate lhs before rhs.
-    if (opcode == BO_Assign) {
-      const uint32_t rhs = doExpr(expr->getRHS());
-      const uint32_t lhs = doExpr(expr->getLHS());
+    const auto *lhsExpr = dyn_cast<HLSLVectorElementExpr>(lhs);
 
-      theBuilder.createStore(lhs, rhs);
-      // Assignment returns a rvalue.
-      return rhs;
+    if (!lhsExpr)
+      return 0;
+
+    if (!isVectorShuffle(lhs)) {
+      // No vector shuffle needed to be generated for this assignment.
+      // Should fall back to the normal handling of assignment.
+      return 0;
     }
 
-    // Try to optimize floatN * float case
-    if (opcode == BO_Mul) {
-      if (const uint32_t result = tryToGenFloatVectorScale(expr))
-        return result;
+    const Expr *base = nullptr;
+    hlsl::VectorMemberAccessPositions accessor;
+    condenseVectorElementExpr(lhsExpr, &base, &accessor);
+
+    const QualType baseType = base->getType();
+    assert(hlsl::IsHLSLVecType(baseType));
+    const auto baseSizse = hlsl::GetHLSLVecSize(baseType);
+
+    llvm::SmallVector<uint32_t, 4> selectors;
+    selectors.resize(baseSizse);
+    // Assume we are selecting all original elements first.
+    for (uint32_t i = 0; i < baseSizse; ++i) {
+      selectors[i] = i;
+    }
+
+    // Now fix up the elements that actually got overwritten by the rhs vector.
+    // Since we are using the rhs vector as the second vector, their index
+    // should be offset'ed by the size of the lhs base vector.
+    for (uint32_t i = 0; i < accessor.Count; ++i) {
+      uint32_t position;
+      accessor.GetPosition(i, &position);
+      selectors[position] = baseSizse + i;
+    }
+
+    const uint32_t baseTypeId = typeTranslator.translateType(baseType);
+    const uint32_t vec1 = doExpr(base);
+    const uint32_t vec1Val = theBuilder.createLoad(baseTypeId, vec1);
+    const uint32_t shuffle =
+        theBuilder.createVectorShuffle(baseTypeId, vec1Val, rhs, selectors);
+
+    theBuilder.createStore(vec1, shuffle);
+
+    // TODO: OK, this return value is incorrect for compound assignments, for
+    // which cases we should return lvalues. Should at least emit errors if
```

Comment:
We are not handling many kind of variables now. This should be a thing to fix when adding support for global variables and more variables with semantics. #102 for this.